File: ui/traits_panel.py
"""Página de Traits."""

import pygame

from config import TOPBAR_H, VIRTUAL_H
from core.traits import TRAITS
from i18n import tr
from theme import (
    ACCENT, ACCENT_HOVER, BAD, BLACK,
    GOOD, GREY, PANEL, PANEL_LIGHT,
    PANEL_LIGHTER, WHITE,
)
from ui.drawing import bake, dim_overlay, draw_panel
from ui.fonts import fit_text, wrap_text


class TraitsPanelMixin:
    """Página de Traits."""

    def close_traits(self):
        self.traits_open = False

    def toggle_traits(self):
        if self.traits_open:
            self.traits_open = False
        else:
            self.close_overlays()      # Traits não convive com Stats nem com Options
            self.traits_open = True
            self.right_panel.close()
            self.left_panel.close()

    def notify_trait_charges(self, n=1):
        """Aviso de carga de trait. Se ganhares mais cargas antes de o aviso desaparecer,
        o mesmo aviso passa a dizer x2, x3... (e volta a contar o tempo)."""
        if n <= 0 or not self.settings.get("trait_notifications", True):
            return
        if self.toast_kind == "trait" and self.toast_timer > 0:
            self.trait_toast_count += n
        else:
            self.trait_toast_count = n
        self.toast_kind = "trait"
        if self.trait_toast_count == 1:
            self.toast_text = tr("You gained a Trait Roll!")
        else:
            self.toast_text = tr("You gained a Trait Roll!  x%d", self.trait_toast_count)
        self.toast_timer = 1.8

    # ---------------------------------------------------------------- página de traits
    def draw_traits_page(self, mouse_pos):
        # Só os botões de navegação (Stats/Options no topo e os laterais) ficam clicáveis por
        # baixo (ver begin_modal); tudo fica escurecido pelo overlay, incluindo os botões laterais.
        self.canvas.blit(dim_overlay(self.vw, VIRTUAL_H, 170), (0, 0))

        # deixa espaço dos dois lados para os botões laterais e os nomes deles (MILESTONES...)
        # e começa por baixo da barra do topo, para não tapar os botões Stats / Options
        panel_w = max(640, min(self.vw - 400, 1080))
        top = TOPBAR_H + 12
        panel_h = VIRTUAL_H - top - 14
        rect = pygame.Rect(self.vw // 2 - panel_w // 2, top, panel_w, panel_h)
        draw_panel(self.canvas, rect, PANEL, radius=16)
        self.register_button(rect, lambda: None, None)     # clicar dentro da página não fecha

        # título + a frase de ajuda por baixo dele (antes ficava por cima e ilegível)
        title = self.font_big.render(tr("Traits"), True, WHITE)
        self.canvas.blit(bake(title, PANEL), (rect.x + 26, rect.y + 16))
        close_rect = pygame.Rect(rect.right - 48, rect.y + 18, 30, 30)
        self.button(close_rect, "X", self.font_small_b, mouse_pos, PANEL_LIGHT, BAD, WHITE,
                    callback=self.close_traits, radius=8)

        sub_y = rect.y + 16 + title.get_height() + 2
        sub_text = fit_text(self.font_small, tr("Your collection — click a trait you own to equip it."),
                            panel_w - 52 - 60)
        self.canvas.blit(bake(self.font_small.render(sub_text, True, GREY), PANEL), (rect.x + 26, sub_y))

        content_top = sub_y + self.font_small.get_height() + 14
        content_h = rect.bottom - 22 - content_top
        list_w = int(panel_w * 0.36)
        list_rect = pygame.Rect(rect.x + 22, content_top, list_w, content_h)
        roll_rect = pygame.Rect(list_rect.right + 22, content_top, panel_w - list_w - 66, content_h)

        self.draw_traits_list(list_rect, mouse_pos)
        self.draw_traits_roll_area(roll_rect, mouse_pos)

    def draw_traits_list(self, rect, mouse_pos):
        self.traits_list_rect = rect

        self.push_clip(rect)
        scroll = self.traits_scroll
        card_w = rect.width - 16          # sobra uma faixa à direita para a barra de scroll
        card_h = 92
        gap = 12
        y = rect.top - scroll

        for i, trait in enumerate(TRAITS):
            crect = pygame.Rect(rect.x, y, card_w, card_h)
            if crect.bottom >= rect.top - 4 and crect.top <= rect.bottom + 4:
                owned = i in self.state.owned_traits
                equipped = self.state.equipped_trait == i
                surf = self.render_trait_card(i, card_w, card_h, owned, equipped)
                # os cartoes estao em cache e ficam por cima do painel (cor lisa): mistura-se
                # cada um com essa cor uma vez e passa a copiar-se sem alfa (ver bake())
                self.canvas.blit(bake(surf, PANEL), crect.topleft)
                if owned:
                    if crect.collidepoint(mouse_pos) and rect.collidepoint(mouse_pos):
                        pygame.draw.rect(self.canvas, WHITE, crect, width=2, border_radius=12)

                    def make_cb(idx=i):
                        return lambda: self.state.equip_trait(idx)
                    self.register_button(crect, make_cb(), "equip")
            y += card_h + gap

        content_h = (len(TRAITS) * (card_h + gap)) - gap
        self.traits_max_scroll = max(0.0, content_h - rect.height)
        self.traits_scroll = max(0.0, min(self.traits_max_scroll, self.traits_scroll))
        self.pop_clip()
        self.draw_scrollbar(rect, scroll, content_h, key="traits", mouse_pos=mouse_pos)

    BATCH_ROW_H = 20

    def traits_batch_height(self):
        """Altura do resumo do 'Use All Charges': cabeçalho + uma linha por trait que saiu."""
        return 36 + len(self.state.last_trait_batch) * self.BATCH_ROW_H + 10

    def draw_traits_batch_summary(self, rect):
        """Mostra o resumo de tudo o que saiu no último 'Usar Todas as Cargas'.
        A caixa adapta-se ao conteúdo (largura e nº de linhas), por isso nada sai dela."""
        batch = self.state.last_trait_batch
        total = sum(batch.values())
        draw_panel(self.canvas, rect, PANEL_LIGHT, radius=12)
        title_txt = fit_text(self.font_small_b, tr("Batch summary (%d charges used):", total), rect.width - 28)
        title = self.font_small_b.render(title_txt, True, WHITE)
        self.canvas.blit(title, (rect.x + 14, rect.y + 10))

        ty = rect.y + 36
        row_h = self.BATCH_ROW_H
        for idx in sorted(batch.keys(), reverse=True):
            trait = TRAITS[idx]
            count = batch[idx]
            swatch = pygame.Rect(rect.x + 14, ty + 3, 14, 14)
            pygame.draw.rect(self.canvas, trait["color"], swatch, border_radius=3)
            line = fit_text(self.font_small, "x%d  %s" % (count, tr(trait["name"])),
                            rect.right - 14 - (swatch.right + 10))
            self.canvas.blit(self.font_small.render(line, True, WHITE), (swatch.right + 10, ty))
            ty += row_h
        return rect

    def draw_traits_roll_area(self, rect, mouse_pos):
        card_w, card_h = 260, 172
        card_rect = pygame.Rect(rect.centerx - card_w // 2, rect.top + 10, card_w, card_h)

        if self.state.last_trait_batch:
            # mais larga que o cartão normal e com altura para TODAS as traits (até 10 linhas)
            sw = max(card_w, min(rect.width - 20, 380))
            summary_rect = pygame.Rect(rect.centerx - sw // 2, rect.top + 10, sw, self.traits_batch_height())
            self.draw_traits_batch_summary(summary_rect)
            y = summary_rect.bottom + 16
        elif self.state.last_trait_roll is not None:
            surf = self.render_trait_card(self.state.last_trait_roll, card_w, card_h,
                                          True, self.state.equipped_trait == self.state.last_trait_roll)
            self.canvas.blit(surf, card_rect.topleft)
            y = card_rect.bottom + 22
        else:
            draw_panel(self.canvas, card_rect, PANEL_LIGHT, radius=12)
            txt = self.font_small.render(tr("You haven't rolled any traits yet."), True, GREY)
            self.canvas.blit(bake(txt, PANEL_LIGHT), txt.get_rect(center=card_rect.center))
            y = card_rect.bottom + 22
        charges = self.state.trait_charges
        charges_txt = self.font_med.render(tr("Charges available: %d", charges), True, GOOD if charges else GREY)
        self.canvas.blit(charges_txt, charges_txt.get_rect(center=(rect.centerx, y + charges_txt.get_height() // 2)))
        y += 34

        chance_str = tr("Every pet roll (manual or Auto Roller) has a %.3f%% chance of giving 1 charge.",
                        self.state.trait_charge_chance() * 100)
        for line in wrap_text(chance_str, self.font_tiny, rect.width - 20):
            chance_txt = self.font_tiny.render(line, True, GREY)
            self.canvas.blit(bake(chance_txt, PANEL), chance_txt.get_rect(center=(rect.centerx, y + chance_txt.get_height() // 2)))
            y += 16
        y += 12

        btn_w = min(260, rect.width - 20)

        def do_single_roll():
            self.state.roll_trait()
            self.state.last_trait_batch = None
            self.play("trait_roll")

        self.button(pygame.Rect(rect.centerx - btn_w // 2, y, btn_w, 46), tr("Roll Trait  (1 charge)"),
                    self.font_med, mouse_pos, ACCENT if charges > 0 else (70, 73, 88),
                    ACCENT_HOVER, BLACK if charges > 0 else GREY,
                    callback=do_single_roll if charges > 0 else None, enabled=charges > 0, radius=10,
                    sfx=None)
        y += 56

        def do_all_rolls():
            n = self.state.trait_charges
            results = {}
            while self.state.trait_charges > 0:
                idx = self.state.roll_trait()
                if idx is not None:
                    results[idx] = results.get(idx, 0) + 1
            if n > 0:
                self.state.last_trait_batch = results
                self.play("trait_roll")
                self.show_toast(tr("Used %d trait charges!", n))

        self.button(pygame.Rect(rect.centerx - btn_w // 2, y, btn_w, 40),
                    tr("Use All Charges (%d)", charges), self.font_small_b, mouse_pos,
                    (52, 120, 80) if charges > 0 else (70, 73, 88), PANEL_LIGHTER,
                    WHITE if charges > 0 else GREY,
                    callback=do_all_rolls if charges > 0 else None, enabled=charges > 0, radius=10,
                    sfx=None)
        y += 54

        # ---- resumo da trait equipada ----
        eq = self.state.equipped_trait
        box_rect = pygame.Rect(rect.x + 10, y, rect.width - 20, rect.bottom - y - 4)
        if box_rect.height > 40:
            draw_panel(self.canvas, box_rect, PANEL_LIGHT, radius=10, shadow=False)
            self.push_clip(box_rect)      # nada do que está lá dentro pode sair da caixa
            inner_w = box_rect.width - 24
            line_h = 20
            if eq is None:
                ty = box_rect.y + 12
                for text in (tr("No trait equipped."),
                             tr("Equip a trait from the list on the side to get its buffs.")):
                    for wline in wrap_text(text, self.font_small, inner_w):
                        self.canvas.blit(bake(self.font_small.render(wline, True, WHITE if ty == box_rect.y + 12 else GREY),
                                              PANEL_LIGHT), (box_rect.x + 12, ty))
                        ty += line_h
            else:
                trait = TRAITS[eq]
                labels = {"money": tr("Money"), "luck": tr("Luck"), "secret_luck": tr("Secret+ Luck"),
                          "mutation": tr("Mutation Chance"), "auto_speed": tr("Auto Speed"),
                          "charge_chance": tr("Trait Charge Chance")}
                items = ["+%.0f%% %s" % (trait["buffs"][k] * 100, label)
                         for k, label in labels.items() if trait["buffs"].get(k)]
                head = self.font_small.render(fit_text(self.font_small, tr("Equipped: %s", tr(trait["name"])), inner_w),
                                              True, WHITE)
                self.canvas.blit(head, (box_rect.x + 12, box_rect.y + 12))
                items_top = box_rect.y + 12 + line_h
                # 1 coluna se couber na altura; senão 2 colunas (as traits raras têm 6-7 buffs)
                one_col_h = len(items) * line_h + 12
                cols = 1 if (items_top - box_rect.y) + one_col_h <= box_rect.height else 2
                col_w = inner_w // cols
                rows = (len(items) + cols - 1) // cols
                for n_i, text in enumerate(items):
                    col, row = n_i // rows, n_i % rows
                    text = fit_text(self.font_small, text, col_w - 8)
                    self.canvas.blit(self.font_small.render(text, True, GREY),
                                     (box_rect.x + 12 + col * col_w, items_top + row * line_h))
            self.pop_clip()

File: src/game/traits_panel.rs
//! Traits page (ui/traits_panel.py).

use super::base::Bo;
use super::{Game, cb};
use crate::config::{TOPBAR_H, VIRTUAL_H};
use crate::core::data::TRAITS;
use crate::core::formatting::format_one_in;
use crate::gfx::{Color, Rect, Surf, draw, ti};
use crate::i18n::tr;
use crate::pyfmt::format as pyformat;
use crate::theme::*;
use crate::ui::cards::render_trait_card;
use crate::ui::drawing::{dim_overlay, draw_panel};
use crate::ui::fonts::{fit_text, wrap_text};
use crate::{args, tr};
use indexmap::IndexMap;
use std::rc::Rc;

const BATCH_ROW_H: i32 = 20;

impl Game {
    pub fn close_traits(&mut self) {
        self.traits_open = false;
    }

    pub fn toggle_traits(&mut self) {
        if self.traits_open {
            self.traits_open = false;
        } else {
            self.close_overlays();
            self.traits_open = true;
            self.right_panel.close();
            self.left_panel.close();
        }
    }

    pub fn notify_trait_charges(&mut self, n: i64) {
        if n <= 0 || !self.settings.get_bool("trait_notifications", true) {
            return;
        }
        if self.toast_kind == Some("trait") && self.toast_timer > 0.0 {
            self.trait_toast_count += n;
        } else {
            self.trait_toast_count = n;
        }
        self.toast_kind = Some("trait");
        self.toast_text = Some(if self.trait_toast_count == 1 { tr("You gained a Trait Roll!") } else { tr!("You gained a Trait Roll!  x%d", self.trait_toast_count) });
        self.toast_timer = 1.8;
    }

    /// Game.render_trait_card: the card with the chance plate filled in.
    pub fn trait_card(&self, trait_index: usize, w: i32, h: i32, owned: bool, equipped: bool) -> Surf {
        let chance_text = format_one_in(self.state.trait_chance(trait_index));
        render_trait_card(trait_index, w, h, owned, equipped, &chance_text)
    }

    pub fn draw_traits_page(&mut self, mouse_pos: (f64, f64)) {
        let ov = dim_overlay(self.vw, VIRTUAL_H, 170);
        self.canvas.blit(&ov, 0, 0);

        let panel_w = 640.max((self.vw - 400).min(1080));
        let top = TOPBAR_H + 12;
        let panel_h = VIRTUAL_H - top - 14;
        let rect = Rect::new(self.vw / 2 - panel_w / 2, top, panel_w, panel_h);
        draw_panel(&mut self.canvas, rect, Some(panel()), 16, true, None);
        self.register_button(rect, Rc::new(|_: &mut Game| {}), None);

        let title = self.f.big.render(&tr("Traits"), WHITE);
        self.canvas.blit(&title, rect.x + 26, rect.y + 16);
        let close_rect = Rect::new(rect.right() - 48, rect.y + 18, 30, 30);
        let sb = self.f.small_b.clone();
        self.button(close_rect, "X", &sb, mouse_pos, panel_light(), BAD, WHITE, cb(|g| g.close_traits()), Bo::r(8));

        let sub_y = rect.y + 16 + title.h + 2;
        let small = self.f.small.clone();
        let sub_text = fit_text(&small, &tr("Your collection — click a trait you own to equip it."), panel_w - 52 - 60);
        let s = small.render(&sub_text, grey());
        self.canvas.blit(&s, rect.x + 26, sub_y);

        let content_top = sub_y + small.get_height() + 14;
        let content_h = rect.bottom() - 22 - content_top;
        let list_w = (panel_w as f64 * 0.36) as i32;
        let list_rect = Rect::new(rect.x + 22, content_top, list_w, content_h);
        let roll_rect = Rect::new(list_rect.right() + 22, content_top, panel_w - list_w - 66, content_h);

        self.draw_traits_list(list_rect, mouse_pos);
        self.draw_traits_roll_area(roll_rect, mouse_pos);
    }

    fn draw_traits_list(&mut self, rect: Rect, mouse_pos: (f64, f64)) {
        self.traits_list_rect = rect;
        self.push_clip(rect);
        let scroll = self.traits_scroll;
        let card_w = rect.w - 16;
        let card_h = 92;
        let gap = 12;
        let mut y = rect.top() as f64 - scroll;
        for i in 0..TRAITS.len() {
            let crect = Rect::new(rect.x, ti(y), card_w, card_h);
            if crect.bottom() >= rect.top() - 4 && crect.top() <= rect.bottom() + 4 {
                let owned = self.state.owned_traits.contains(&i);
                let equipped = self.state.equipped_trait == Some(i);
                let surf = self.trait_card(i, card_w, card_h, owned, equipped);
                self.canvas.blit(&surf, crect.x, crect.y);
                if owned {
                    if crect.collidepoint(mouse_pos) && rect.collidepoint(mouse_pos) {
                        draw::rect(&mut self.canvas, WHITE, crect, 2, 12);
                    }
                    self.register_button(
                        crect,
                        Rc::new(move |g: &mut Game| {
                            g.state.equip_trait(i);
                        }),
                        Some("equip"),
                    );
                }
            }
            y += (card_h + gap) as f64;
        }
        let content_h = (TRAITS.len() as i32 * (card_h + gap) - gap) as f64;
        self.traits_max_scroll = (content_h - rect.h as f64).max(0.0);
        self.traits_scroll = self.traits_scroll.min(self.traits_max_scroll).max(0.0);
        self.pop_clip();
        self.draw_scrollbar(rect, scroll, content_h, Some("traits"), Some(mouse_pos));
    }

    fn traits_batch_height(&self) -> i32 {
        36 + self.state.last_trait_batch.as_ref().map(|b| b.len() as i32).unwrap_or(0) * BATCH_ROW_H + 10
    }

    fn draw_traits_batch_summary(&mut self, rect: Rect) {
        let batch = self.state.last_trait_batch.clone().unwrap_or_default();
        let total: i64 = batch.values().sum();
        draw_panel(&mut self.canvas, rect, Some(panel_light()), 12, true, None);
        let sb = self.f.small_b.clone();
        let small = self.f.small.clone();
        let title_txt = fit_text(&sb, &tr!("Batch summary (%d charges used):", total), rect.w - 28);
        let title = sb.render(&title_txt, WHITE);
        self.canvas.blit(&title, rect.x + 14, rect.y + 10);
        let mut ty = rect.y + 36;
        let mut keys: Vec<usize> = batch.keys().copied().collect();
        keys.sort_unstable_by(|a, b| b.cmp(a));
        for idx in keys {
            let t = &TRAITS[idx];
            let count = batch[&idx];
            let swatch = Rect::new(rect.x + 14, ty + 3, 14, 14);
            draw::rect(&mut self.canvas, t.color, swatch, 0, 3);
            let line = fit_text(&small, &pyformat("x%d  %s", &args![count, tr(t.name)]), rect.right() - 14 - (swatch.right() + 10));
            let s = small.render(&line, WHITE);
            self.canvas.blit(&s, swatch.right() + 10, ty);
            ty += BATCH_ROW_H;
        }
    }

    fn draw_traits_roll_area(&mut self, rect: Rect, mouse_pos: (f64, f64)) {
        let (card_w, card_h) = (260, 172);
        let card_rect = Rect::new(rect.centerx() - card_w / 2, rect.top() + 10, card_w, card_h);
        let small = self.f.small.clone();
        let med = self.f.med.clone();
        let tiny = self.f.tiny.clone();
        let sb = self.f.small_b.clone();

        let has_batch = self.state.last_trait_batch.as_ref().is_some_and(|b| !b.is_empty());
        let mut y;
        if has_batch {
            let sw = card_w.max((rect.w - 20).min(380));
            let summary_rect = Rect::new(rect.centerx() - sw / 2, rect.top() + 10, sw, self.traits_batch_height());
            self.draw_traits_batch_summary(summary_rect);
            y = summary_rect.bottom() + 16;
        } else if let Some(last) = self.state.last_trait_roll {
            let surf = self.trait_card(last, card_w, card_h, true, self.state.equipped_trait == Some(last));
            self.canvas.blit(&surf, card_rect.x, card_rect.y);
            y = card_rect.bottom() + 22;
        } else {
            draw_panel(&mut self.canvas, card_rect, Some(panel_light()), 12, true, None);
            let txt = small.render(&tr("You haven't rolled any traits yet."), grey());
            let r = Rect::with_center(txt.w, txt.h, card_rect.center());
            self.canvas.blit(&txt, r.x, r.y);
            y = card_rect.bottom() + 22;
        }
        let charges = self.state.trait_charges;
        let charges_txt = med.render(&tr!("Charges available: %d", charges), if charges != 0 { GOOD } else { grey() });
        let r = Rect::with_center(charges_txt.w, charges_txt.h, (rect.centerx(), y + charges_txt.h / 2));
        self.canvas.blit(&charges_txt, r.x, r.y);
        y += 34;

        let chance_str = tr!("Every pet roll (manual or Auto Roller) has a %.3f%% chance of giving 1 charge.", self.state.trait_charge_chance() * 100.0);
        for line in wrap_text(&chance_str, &tiny, rect.w - 20) {
            let t = tiny.render(&line, grey());
            let r = Rect::with_center(t.w, t.h, (rect.centerx(), y + t.h / 2));
            self.canvas.blit(&t, r.x, r.y);
            y += 16;
        }
        y += 12;

        let btn_w = 260.min(rect.w - 20);
        let can = charges > 0;
        self.button(
            Rect::new(rect.centerx() - btn_w / 2, y, btn_w, 46),
            &tr("Roll Trait  (1 charge)"),
            &med,
            mouse_pos,
            if can { accent() } else { Color::rgb(70, 73, 88) },
            accent_hover(),
            if can { BLACK } else { grey() },
            if can {
                cb(|g| {
                    g.state.roll_trait();
                    g.state.last_trait_batch = None;
                    g.play("trait_roll", 0.0);
                })
            } else {
                None
            },
            Bo::r(10).enabled(can).sfx(None),
        );
        y += 56;

        self.button(
            Rect::new(rect.centerx() - btn_w / 2, y, btn_w, 40),
            &tr!("Use All Charges (%d)", charges),
            &sb,
            mouse_pos,
            if can { Color::rgb(52, 120, 80) } else { Color::rgb(70, 73, 88) },
            panel_lighter(),
            if can { WHITE } else { grey() },
            if can {
                cb(|g| {
                    let n = g.state.trait_charges;
                    let mut results: IndexMap<usize, i64> = IndexMap::new();
                    while g.state.trait_charges > 0 {
                        if let Some(idx) = g.state.roll_trait() {
                            *results.entry(idx).or_insert(0) += 1;
                        }
                    }
                    if n > 0 {
                        g.state.last_trait_batch = Some(results);
                        g.play("trait_roll", 0.0);
                        g.show_toast(&tr!("Used %d trait charges!", n), 1.8);
                    }
                })
            } else {
                None
            },
            Bo::r(10).enabled(can).sfx(None),
        );
        y += 54;

        let eq = self.state.equipped_trait;
        let box_rect = Rect::new(rect.x + 10, y, rect.w - 20, rect.bottom() - y - 4);
        if box_rect.h > 40 {
            draw_panel(&mut self.canvas, box_rect, Some(panel_light()), 10, false, None);
            self.push_clip(box_rect);
            let inner_w = box_rect.w - 24;
            let line_h = 20;
            match eq {
                None => {
                    let mut ty = box_rect.y + 12;
                    for text in [tr("No trait equipped."), tr("Equip a trait from the list on the side to get its buffs.")] {
                        for wline in wrap_text(&text, &small, inner_w) {
                            let s = small.render(&wline, if ty == box_rect.y + 12 { WHITE } else { grey() });
                            self.canvas.blit(&s, box_rect.x + 12, ty);
                            ty += line_h;
                        }
                    }
                }
                Some(eq) => {
                    let t = &TRAITS[eq];
                    let labels = [
                        ("money", tr("Money")),
                        ("luck", tr("Luck")),
                        ("secret_luck", tr("Secret+ Luck")),
                        ("mutation", tr("Mutation Chance")),
                        ("auto_speed", tr("Auto Speed")),
                        ("charge_chance", tr("Trait Charge Chance")),
                    ];
                    let items: Vec<String> = labels
                        .iter()
                        .filter(|(k, _)| t.buff(k) != 0.0)
                        .map(|(k, label)| pyformat("+%.0f%% %s", &args![t.buff(k) * 100.0, label]))
                        .collect();
                    let head = small.render(&fit_text(&small, &tr!("Equipped: %s", tr(t.name)), inner_w), WHITE);
                    self.canvas.blit(&head, box_rect.x + 12, box_rect.y + 12);
                    let items_top = box_rect.y + 12 + line_h;
                    let one_col_h = items.len() as i32 * line_h + 12;
                    let cols = if (items_top - box_rect.y) + one_col_h <= box_rect.h { 1 } else { 2 };
                    let col_w = inner_w / cols;
                    let rows = (items.len() as i32 + cols - 1) / cols;
                    for (n_i, text) in items.iter().enumerate() {
                        let n_i = n_i as i32;
                        let (col, row) = (n_i / rows, n_i % rows);
                        let text = fit_text(&small, text, col_w - 8);
                        let s = small.render(&text, grey());
                        self.canvas.blit(&s, box_rect.x + 12 + col * col_w, items_top + row * line_h);
                    }
                }
            }
            self.pop_clip();
        }
    }
}
